[CMake] Add functorch target

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83464
* #83777
* #84788

Move functorch/functorch into `functorch` folder
- Add functorch/CMakeLists.txt that adds `functorch` native python exension
- Modify `setup.py` to package pytorch and functorch together into a single wheel
- Modify `functorch.__version__` is not equal to that of `torch.__version__`
- Add dummy `functorch/setup.py` file for the projects that still want to build it

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

@diff-train-skip-merge Manually copied this commit to fbsync to fix shipit

malfet committed:
[CMake] Add functorch target

Add libfunctorch target as PyTorch cmake build system

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

[ghstack-poisoned]

zou3519 commented on setup.py:
```diff
@@ -614,6 +614,19 @@ def build_extensions(self):
                     os.makedirs(dst_dir)
                 self.copy_file(src, dst)
                 i += 1
+
```
Just to check my understanding: after this PR, we will no longer have a standalone functorch build. Is that correct? (That isn't a problem or requirement, though it has some implications for the transition plan).

Users are still going to install functorch from source for a while (we have folks who run functorch in their CIs). For those people, we should make it so that installing functorch from the original location (https://github.com/pytorch/functorch) does something sane. Today there's a shipit sync from the functorch in Pytorch to https://github.com/pytorch/functorch to make it sane.

So here's the order I think we should do things in:
1. Turn off the shipit sync
2. Move functorch python code to a `torch._functorch` folder
3. Merge this PR
4. Change the pytorch/functorch repo to have a package that ends up doing `from torch._functorch import *`. If someone installs both PyTorch and `https://github.com/pytorch/functorch`, then `https://github.com/pytorch/functorch` will take precedence, but it will import the correct thing.

What do you think? Is there something easier we can do? If not, then I can handle 1, 2, and 4 (2 might take a week or so because we need to get some code reviews for it)

zou3519 replied:
@albanD suggested that instead of step (1), we can distribute a `_functorch_pytorch` and a `functorch` package with PyTorch. The `_functorch_pytorch` package would contain the extension, and the `functorch` package would just `from _functorch_pytorch import *`. This would avoid us needing to wait to finish code review for step 1.

zou3519 replied:
Nikita and I discussed more offline, we can just turn pytorch/functorch into a dummy package that doesn't contain anything. pytorch/functorch's setup.py lists `torch` as a dependency so that users will get `import functorch` automatically if they install pytorch/functorch from source (or if they install regular PyTorch)

zou3519 reviewed (commented):
This is awesome! I have two high-level comments for things we should resolve before merging this:
- the transition plan (see comment above), we want to make sure that installing https://github.com/pytorch/functorch still works for a period of time for BC
- fbcode. Does this work internally?

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

[ghstack-poisoned]

zou3519 commented:
I see
```
ImportError: cannot import name '_C' from 'functorch' (/opt/conda/lib/python3.7/site-packages/functorch/__init__.py)
```
I'm getting the same thing locally when I try `python setup.py install`. `python setup.py develop` works though.

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39058811).

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

[ghstack-poisoned]

malfet commented:
> I see
> 
> ```
> ImportError: cannot import name '_C' from 'functorch' (/opt/conda/lib/python3.7/site-packages/functorch/__init__.py)
> ```
> 
> in CI: https://github.com/pytorch/pytorch/runs/8023231814?check_suite_focus=true I'm getting the same thing locally when I try `python setup.py install`. `python setup.py develop` works though.

Fixed that last week, both workflow should work fine now.

malfet commented:
Windows failures are somewhat interesting:
 - functorch (and libtorch_python) were definitely compiled against python-3.9, but then in runtime it complains about python-3.10 incompatibility...

And test_dims segfauls on some configurations with null pointer dereference, while running ` _parse_test(2, 0, "x", "y", b="y")`

zou3519 commented:
> definitely compiled against python-3.9, but then in runtime it complains about python-3.10 incompatibility...

Maybe related: I hit a problem when I was testing this PR two weeks ago, where I compiled this PR against python-3.8 but it complained about compatibility with python 3.9 during runtime. I originally brushed this off as a problem with my environment, but it's sounding similar

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

[ghstack-poisoned]

malfet commented:
segfault seems to be caused by spurious dependency on `libpython.so`, which happens only if cmake-3.13.5 is used. Trying to workaround by using `add_library` rather than `Python3_add_library`...

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

[ghstack-poisoned]

malfet commented:
Looks like the only issue that remains are crashes, when build with older cmake. Posting backtrace here:
```
Thread 1 "python" received signal SIGSEGV, Segmentation fault.
convertitem () at /home/builder/tkoch/workspace/python_1648536129212/work/Python/getargs.c:598
598	/home/builder/tkoch/workspace/python_1648536129212/work/Python/getargs.c: No such file or directory.
(gdb) bt
#0  convertitem () at /home/builder/tkoch/workspace/python_1648536129212/work/Python/getargs.c:598
#1  0x000055b263fc433f in vgetargskeywordsfast_impl () at /home/builder/tkoch/workspace/python_1648536129212/work/Python/getargs.c:2152
#2  0x000055b263fc5d4a in _PyArg_ParseStackAndKeywords_SizeT (args=<optimized out>, nargs=<optimized out>, kwnames=<optimized out>, parser=<optimized out>)
    at /home/builder/tkoch/workspace/python_1648536129212/work/Python/getargs.c:1558
#3  0x00007fc71b01f948 in py::vector_args::parse(char const*, std::initializer_list<char const*>, std::initializer_list<py::handle*>, int, int)::{lambda()#1}::operator()() const (this=0x7fff93562998) at ../functorch/csrc/dim/minpybind.h:620
#4  0x00007fc71b01f686 in py::vector_args::parse (this=0x7fff93562b38, fname_cstr=0x7fc71b0553c7 "_parse_test", names=..., values=..., required=2, kwonly=0)
    at ../functorch/csrc/dim/minpybind.h:668
#5  0x00007fc71b00b46c in _parse_test (self_=0x7fc71b358890, args=0x7fc71a36c1d0, nargs=4, kwnames=0x7fca56721510) at ../functorch/csrc/dim/dim.cpp:3089
#6  0x000055b263f7b6ec in _PyMethodDef_RawFastCallKeywords () at /home/builder/tkoch/workspace/python_1648536129212/work/Objects/call.c:659
```

zou3519 commented on functorch/setup.py:
```diff
@@ -1,149 +0,0 @@
-# Copyright (c) Facebook, Inc. and its affiliates.
-# All rights reserved.
-#
```
We should have a dummy functorch/setup.py that is a one-liner that raises a warning that one does not need to use it anymore to smooth the transition. PyTorch devs will cd into functorch and attempt to run setup.py develop.

Also, it might be good to mention that the previous functorch installation needs to get uninstalled (I discovered this while running through https://docs.google.com/document/d/1nMiUkuewG0RfRMs2gNxr5VKBDlCm8xAgF8FrRLFw4lU/edit# )

zou3519 commented on functorch/setup.py:
```diff
@@ -1,149 +0,0 @@
-# This source code is licensed under the BSD-style license found in the
-# LICENSE file in the root directory of this source tree.
-
-import distutils.command.clean
-import shutil
```
We've got a UX problem where the way dynamo installs functorch no longer works (https://github.com/pytorch/torchdynamo/blob/d88a0ac0e4a799617e59a05fec79d329781f3ea7/Makefile#L54-L55 ) due to setup.py not being legit anymore. I'm not sure how to resolve this: maybe we should just have a trivial setup.py?

zou3519 replied:
There are at least three users for this: https://github.com/search?q=subdirectory%3Dfunctorch&type=code . To err on the safe side, we should have a dummy setup.py that does invoke setuptools

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

[ghstack-poisoned]

zou3519 commented on functorch/setup.py:
```diff
@@ -1,149 +1,4 @@
-            "build_ext": BuildExtension_.with_options(no_python_abi_suffix=True),
-            'clean': clean,
-        })
+# Welcome to the FuncTorch setup.py.
+# As it has been embedded into PyTorch it's not supposed to do anything
+if __name__ == "__main__":
+    print("Mostly harmless")
```
@malfet Do you think we could (or should) change this to https://github.com/pytorch/functorch/blob/e108ed74228ef4db8200b1df59138b1696ad573e/setup.py ?

The context is, there are at least three projects that install functorch via the following command: https://github.com/pytorch/torchdynamo/blob/d88a0ac0e4a799617e59a05fec79d329781f3ea7/Makefile#L54-L55 . It would be best if we don't hard-break them and I can migrate them off one-by-one after we do the merge

malfet replied:
Sure, let me update it like that.

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Add libfunctorch target as PyTorch cmake build system

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

[ghstack-poisoned]

malfet committed:
Update on "[CMake] Add functorch target"


Move functorch/functorch into `functorch` folder
- Add functorch/CMakeLists.txt that adds `functorch` native python exension
- Modify `setup.py` to package pytorch and functorch together into a single wheel
- Modify `functorch.__version__` is not equal to that of `torch.__version__`
- Add dummy `funcotrch/setup.py` file for the projects that still want to build it

Differential Revision: [D39058811](https://our.internmc.facebook.com/intern/diff/D39058811)

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39058811).

malfet commented:
@pytorchbot merge -f "I've seen it green"

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39058811).
